### ISSUE TITLE
Teleport Concurrent SSH Batching

### ISSUE THREAD

**Issue by gwellington** [gravitational/teleport#10241]
### What
What would you like Teleport to do differently?
Teleport clients should be server aware of the max number of sessions allowed.  When a server imposes a limit and the client requests SSH connections to more than the number of allowed, the client should batch sessions and queue them accordingly.

For instance, if 10 connections are allowed, then the client should queue up 10 sessions in parallel.  As each session completes, the client should queue up N more sessions until all hosts have been accessed.   Failure of 1 host should NOT preclude other hosts from being accessed.

### How
How would you implement this?

Naive solution:
Implement a `bucket` or `batch` argument to `tsh ssh` which takes in the max number of concurrent sessions to run at once and queue up additional connections as others finish.

More Awesome solution:
Client should be smart enough to ask the server what my limit is and use that as the batch size.

Even More Awesome Solution:
Client should be server-limit aware but also allow users to pass in their own batch size (regardless of server limit).  A warning might be nice if it exceeds it.

### Why
Why do you need this?

We need to operate in an environment which limits the number of concurrent SSH connections and need to manage a large fleet.   The current implementation of this makes the parallel SSH feature unusable. :( 

### Workaround
If a workaround exists, please include it.

No work around other than writing our own implementation of `tsh ssh`

**Comment by kazimierzbudzyk** [comment]
@russjones for awareness - we discussed this few months back. We tried running proxies with in-memory mode enabled (as you suggested) but unfortunately it didn't improve the issue of unreliable session reclaim.

**Comment by kazimierzbudzyk** [comment]
Spinning up separate issue for the slow session reclaim problem https://github.com/gravitational/teleport/issues/10363.

**Comment by russjones** [comment]
Customer is trying to use label based SSH access to run commands across a batch of 100+ machines but concurrent session control only allows 10 at a time.

**Comment by klizhentas** [comment]
Sounds good to me. I think we discussed this. You can solve that with a simple semaphore on `tsh ssh` and `tsh ssh proxy` side that can hold extra commands until the connection pool will get a new slot available.
